ECSW Hyperreduction Code and Tests

biondicc opened the pull request:
Pull request for the addition of the energy-conserving sampling and weighting hyperreduction technique into the projection-based ROMs previously tested.

Non-negative least squares solver (added to the linear_solver directory):
- Implemented the NNLS algorithm developed by Lawson and Hanson (presented in _Solving Least Squares Problems_ [https://epubs.siam.org/doi/10.1137/1.9781611971217](url) )
- The NNLS solver can be modified via the inputs to the class:
- By default the internal loop least squares problem is solved directly using the Normal equations, by setting iter_solver_ to true an iterative solver can be used instead
- By default the exit criterion is the update to the original Lawson algorithm presented by Chapman et al. in _Accelerated mesh sampling for the hyper reduction of nonlinear computational models_ [https://doi.org/10.1002/nme.5332](url). This can be changed back to the original gradient based exit condition presented by Lawson.
- A helper file was added which contains multiple functions for changing between Epetra and Eigen data structures.

ode_solver directory changes:
- Added hyperreduced PG solver, following the structure of the other ROM solvers with new functions to find the hyperreduced residual and Jacobian

parameters directory changes:
- Renamed one of the POD tests from old ROM code (POD_adaptive_sampling -> POD_adaptive_sampling_run)
- Added parameters_hyper_reduction.cpp to control hyperreduction and NNLS solver settings

reduced_order directory changes:
- Moved pod_adaptive_sampling.cpp/h from the tests folder and made it into a class
- Created a base class for which adaptive sampling with and without hyperreduction are built on top of
- Added code to assemble the training data for the NNLS problem, using a base class and two derived classes for the residual and Jacobian-based ECSW approaches
-  A third sampling procedure class (hyper_reduced_sampling_error_updated.cpp) is implemented with the new DWR error indicator which depends on the new test location class (hrom_test_location.cpp)

Test changes:
- pod_adaptive_sampling.cpp was moved into a class, and the pod_adaptive_sampling_run.cpp test can be used to build an instance of this class and run the sampling procedure to the tolerance set in the parameter file
-  hyper_adaptive_sampling_run.cpp and hyper_adaptive_sampling_new_error.cpp perform similar tests running the sampling procedure with hyperreduction in both and the updated DWR error in the latter
- hyper_reduction_post_sampling.cpp will conduct hyperreduction after the sampling procedure has been completed, and hyper_reduction_comparison.cpp will do the same and compare it to a similar ROM without hyperreduction
- build_NNLS_problem.cpp will test the assembly and results of a sample NNLS problem built from a ROM problem
- ROM_error_post_sampling.cpp and HROM_error_post_sampling will find the "true" error at 400 evenly distributed points for the NACA 0012 transonic 2 parameter test case for a ROM without and with hyperreduction, respectively

jbrillon reviewed (approved):
LGTM

cpethrick commented on tests/unit_tests/CMakeLists.txt:
```diff
@@ -11,3 +11,4 @@ add_subdirectory(optimization)
 add_subdirectory(operator_tests)
 add_subdirectory(flow_variable_tests)
 add_subdirectory(ode_solver_unit_test)
+add_subdirectory(linear_solver)
```
Would it be more appropriate to specify NNLS solver in the folder name?

biondicc replied:
I kept is as linear solver just in case over solvers get added to the linear solver folder in the source code, and then there unit tests can be added here

cpethrick replied:
Sounds good.

cpethrick reviewed (commented):
Great work adding HROM to the code! It's a well-done PR, and the structure of your new code looks good.

I've made several comments about readibility, which should be pretty quick to resolve. Also, can you look over the files and make sure that debugging `cout` statements are removed to ensure that the code is tidy? I've made a few comments to make it easier to find these, but there may be other places.

I'm not too familiar with the ROM field, but it may be helpful to add more literature references where suitable, such as if a function is implementing something exactly per some equation in a paper. You've already done this in a few places, but I'm just pointing it out in case you can think of any other specific places where references could be added.

I'm happy to approve once you have responded to my comments and posted ctest results! Great work :)

biondicc commented on src/linear_solver/NNLS_solver.h:
```diff
@@ -0,0 +1,161 @@
+/* Non-Negagive Least Squares Solver for Epetra Structures
+ * Based on the NNLS in Eigen unsupported and in MATLAB:
+ * https://gitlab.com/libeigen/eigen/-/blob/master/unsupported/Eigen/NNLS
```
Copyrights added at the top of the new commit

cpethrick replied:
Looks good

biondicc commented on src/ode_solver/hyper_reduced_petrov_galerkin_ode_solver.cpp:
```diff
@@ -0,0 +1,473 @@
+    // Find test basis W with hyper-reduced Jacobian
+    const Epetra_CrsMatrix epetra_pod_basis = pod->getPODBasis()->trilinos_matrix();
+    std::shared_ptr<Epetra_CrsMatrix> epetra_test_basis = generate_test_basis(*reduced_system_matrix, epetra_pod_basis);
+    // this->pcout << "Test Basis" << std::endl;
```
Removed

biondicc commented on src/ode_solver/hyper_reduced_petrov_galerkin_ode_solver.h:
```diff
@@ -0,0 +1,57 @@
+    std::shared_ptr<ProperOrthogonalDecomposition::PODBase<dim>> pod;
+
+    /// ECSW hyper-reduction weights
+    Epetra_Vector xi;
```
Changed to ECSW_weights in new commit

cpethrick replied:
Thanks for changing! Can you check through other files to make sure that this change is applied consistently in those other files? I pointed one out in another comment, but maybe there are others too.

biondicc replied:
I think it was only still xi in that one other file, so should all be ECSW_weights now

biondicc commented on src/ode_solver/ode_solver_factory.cpp:
```diff
@@ -49,6 +50,21 @@ std::shared_ptr<ODESolverBase<dim,real,MeshType>> ODESolverFactory<dim,real,Mesh
+    if(ode_solver_type == ODEEnum::hyper_reduced_petrov_galerkin_solver) 
+        return std::make_shared<HyperReducedODESolver<dim,real,MeshType>>(dg_input, pod, weights);
+    else {
+        display_error_ode_solver_factory(ode_solver_type, true);
```
Added on line 126 in new commit

cpethrick replied:
Thanks!

biondicc commented on src/parameters/parameters_hyper_reduction.cpp:
```diff
@@ -0,0 +1,52 @@
+        prm.declare_entry("NNLS_tol", "1E-6",
+                          dealii::Patterns::Double(0, dealii::Patterns::Double::max_double_value),
+                          "Tolerance for the NNLS solver");
+        prm.declare_entry("NNLS_max_iter", "5000",
```
Going to keep them separate as both a linear solver and the NNLS solver can be used in the same sampling procedure for different purposes/run to different residual tolerances

biondicc commented on src/reduced_order/adaptive_sampling_base.cpp:
```diff
@@ -0,0 +1,357 @@
+    MatrixXd parameters(n_rows, snapshot_parameters.cols());
+    VectorXd errors(n_rows);
+
+    int i;
```
Added a comment to the code. Will also note the purpose here:
-The first for loop runs through the FOM snapshots and adds each location to the parameters matrix as well as a zero error value in the errors vector (error between the FOM and ROM is zero at the snapshot locations)
-The second for loop runs through the ROM points and adds each location to the parameters matrix and adds the total error value at the ROM point to the errors vector
Both the ROM points and FOM snapshot locations are used to build the RBF

cpethrick replied:
That's more clear. Thanks for adding the comments!

biondicc commented on src/ode_solver/hyper_reduced_petrov_galerkin_ode_solver.h:
```diff
@@ -0,0 +1,57 @@
+{
+public:
+    /// Default constructor that will set the constants.
+    HyperReducedODESolver(std::shared_ptr< DGBase<dim, real, MeshType> > dg_input, std::shared_ptr<ProperOrthogonalDecomposition::PODBase<dim>> pod, Epetra_Vector weights); ///< Constructor.
```
Removed

biondicc commented on src/ode_solver/hyper_reduced_petrov_galerkin_ode_solver.cpp:
```diff
@@ -0,0 +1,473 @@
+template class HyperReducedODESolver<PHILIP_DIM, double, dealii::Triangulation<PHILIP_DIM>>;
+template class HyperReducedODESolver<PHILIP_DIM, double, dealii::parallel::shared::Triangulation<PHILIP_DIM>>;
+#if PHILIP_DIM != 1
+template class HyperReducedODESolver<PHILIP_DIM, double, dealii::parallel::distributed::Triangulation<PHILIP_DIM>>;
```
Added indent

biondicc commented on src/reduced_order/adaptive_sampling_base.cpp:
```diff
@@ -0,0 +1,357 @@
+
+    // Set parameters.
+    ROL::ParameterList parlist;
+    //parlist.sublist("General").set("Recompute Objective Function",false);
```
removed

biondicc commented on src/reduced_order/assemble_ECSW_jacobian.cpp:
```diff
@@ -0,0 +1,238 @@
+
+    // Get dimensions of the problem
+    int num_snaps = snapshotMatrix.cols(); // Number of snapshots used to build the POD basis
+    int n = epetra_pod_basis.NumGlobalCols(); // Reduced subspace dimension
```
Changed to the following:
num_snaps_POD
n_reduced_dim_POD
N_FOM_dim
num_elements_N_e

cpethrick replied:
Thanks, those are more descriptive!

biondicc commented on src/reduced_order/assemble_ECSW_jacobian.cpp:
```diff
@@ -0,0 +1,238 @@
+
+    C.FillComplete(ColMap, RowMap);
+
+    // std::cout << " Matrix C "<< std::endl;
```
removed

biondicc commented on src/reduced_order/hyper_reduced_sampling_error_updated.cpp:
```diff
@@ -35,66 +38,147 @@ AdaptiveSampling<dim, nstate>::AdaptiveSampling(const PHiLiP::Parameters::AllPar
         max_error_params = getMaxErrorROM();
 
-        this->pcout << "Max error is: " << max_error << std::endl;
+        std::cout << "Max error is: " << max_error << std::endl;
```
Changed it back to using pcout

biondicc commented on src/testing/hyper_reduction_comparison.cpp:
```diff
@@ -0,0 +1,264 @@
+
+
+template <int dim, int nstate>
+bool HyperReductionComparison<dim, nstate>::getSnapshotParamsFromFile() const{
```
Moved the repeated functions in a couple tests to a helper function file called rom_import_helper_functions

ssaishruthi98 commented on src/linear_solver/NNLS_solver.cpp:
```diff
@@ -0,0 +1,364 @@
+  Epetra_Vector grad_col (A_.ColMap());
+
+  // OUTER LOOP
+  while(true){
```
Would it be possible to use exit flags for the outer and inner loop? Having clear exit conditions might help future users understand the code easier. If exit flags aren't a viable option for the actions being performed here, then there's no need to change it! 😄

biondicc replied:
Modified the inner loop to use a flag, but left the outer the same due to the use of multiple return statements

ssaishruthi98 reviewed (commented):
Congrats on your first PR! It's structured very well and is easy to understand.

I've left a few minor comments. I'm good to approve after the comments are addressed and ctest results are posted! 😄

biondicc commented on src/ode_solver/hyper_reduced_petrov_galerkin_ode_solver.cpp:
```diff
@@ -0,0 +1,473 @@
+void HyperReducedODESolver<dim,real,MeshType>::allocate_ode_system ()
+{
+    /*Projection of initial conditions on reduced-order subspace, refer to Equation 19 in:
+    Washabaugh, K. M., Zahr, M. J., & Farhat, C. (2016).
```
added references to multiple header files

cpethrick replied:
Great! That will be very helpful for future students :)

biondicc commented on src/ode_solver/hyper_reduced_petrov_galerkin_ode_solver.cpp:
```diff
@@ -0,0 +1,473 @@
+template <int dim, typename real, typename MeshType>
+std::shared_ptr<Epetra_CrsMatrix> HyperReducedODESolver<dim,real,MeshType>::generate_hyper_reduced_jacobian(const Epetra_CrsMatrix &system_matrix)
+{
+    // NOTE: ASSUMING L_e matrix is equal to the L_e_plus matrix (i.e. no neighbouring elements considered in stencil)
```
Added comment with equation # and reference

cpethrick replied:
Great! The new comment is quite helpful :)

biondicc commented on src/ode_solver/hyper_reduced_petrov_galerkin_ode_solver.cpp:
```diff
@@ -0,0 +1,471 @@
+#include "dg/dg_base.hpp"
+#include "ode_solver_base.h"
+#include "linear_solver/linear_solver.h"
+#include "reduced_order/pod_basis_base.h"
```
removed

biondicc commented on src/reduced_order/assemble_ECSW_jacobian.cpp:
```diff
@@ -0,0 +1,233 @@
+#include <Epetra_CrsMatrix.h>
+#include <Epetra_Map.h>
+#include <Epetra_Vector.h>
+#include <EpetraExt_MatrixMatrix.h>
```
removed

biondicc commented on src/reduced_order/assemble_ECSW_residual.cpp:
```diff
@@ -0,0 +1,167 @@
+#include <eigen/Eigen/Dense>
+#include <Epetra_CrsMatrix.h>
+#include <Epetra_Map.h>
+#include <Epetra_Vector.h>
```
removed

biondicc commented on src/reduced_order/assemble_ECSW_training_data_base.cpp:
```diff
@@ -0,0 +1,119 @@
+#include <Epetra_CrsMatrix.h>
+#include <Epetra_Map.h>
+#include <Epetra_Vector.h>
+#include <EpetraExt_MatrixMatrix.h>
```
you're good :)

cpethrick commented on src/ode_solver/hyper_reduced_petrov_galerkin_ode_solver.h:
```diff
@@ -0,0 +1,81 @@
+    /// Default constructor that will set the constants.
+    HyperReducedODESolver(std::shared_ptr< DGBase<dim, real, MeshType> > dg_input, std::shared_ptr<ProperOrthogonalDecomposition::PODBase<dim>> pod, Epetra_Vector weights);
+
+    ///POD
```
```suggestion
    /// POD
```

biondicc replied:
added

cpethrick commented on src/ode_solver/ode_solver_factory.cpp:
```diff
@@ -93,6 +123,7 @@ void ODESolverFactory<dim,real,MeshType>::display_error_ode_solver_factory(Param
     else if (ode_solver_type == ODEEnum::rrk_explicit_solver)           solver_string = "rrk_explicit";
     else if (ode_solver_type == ODEEnum::pod_galerkin_solver)           solver_string = "pod_galerkin";
     else if (ode_solver_type == ODEEnum::pod_petrov_galerkin_solver)    solver_string = "pod_petrov_galerkin";
+    else if (ode_solver_type == ODEEnum::hyper_reduced_petrov_galerkin_solver)    solver_string = "hyper_reduced_petrov_galerkin";
```
Sorry for being a bit picky, but can you put `solver_string = '...'` on the next line so that it lines up with the other `solver_string`s?

biondicc replied:
done and modified the line below as well to line up

cpethrick replied:
Thank you :)

cpethrick commented on src/reduced_order/assemble_ECSW_residual.cpp:
```diff
@@ -0,0 +1,162 @@
+}
+
+#if PHILIP_DIM==1
+        template class AssembleECSWRes<PHILIP_DIM, PHILIP_DIM>;
```
```suggestion
    template class AssembleECSWRes<PHILIP_DIM, PHILIP_DIM>;
```

biondicc replied:
removed extra tab

cpethrick commented on src/reduced_order/assemble_ECSW_residual.cpp:
```diff
@@ -0,0 +1,162 @@
+#endif
+
+#if PHILIP_DIM!=1
+        template class AssembleECSWRes<PHILIP_DIM, PHILIP_DIM+2>;
```
```suggestion
    template class AssembleECSWRes<PHILIP_DIM, PHILIP_DIM+2>;
```

biondicc replied:
removed extra tab
removed in assemble_ECSW_jacobian.cpp and assemble_ECSW_training_base.cpp as well

cpethrick commented on src/reduced_order/pod_adaptive_sampling.cpp:
```diff
@@ -0,0 +1,189 @@
+        }
+
+        if(element == this->rom_locations.end() && snapshot_exists == false){
+            //ProperOrthogonalDecomposition::ROMSolution<dim, nstate> rom_solution = solveSnapshotROM(midpoint);
```
```suggestion

```

biondicc replied:
removed

cpethrick commented on src/reduced_order/pod_adaptive_sampling.cpp:
```diff
@@ -0,0 +1,189 @@
+    // Solve implicit solution
+    auto ode_solver_type = Parameters::ODESolverParam::ODESolverEnum::pod_petrov_galerkin_solver;
+    flow_solver->ode_solver =  PHiLiP::ODE::ODESolverFactory<dim, double>::create_ODESolver_manual(ode_solver_type, flow_solver->dg, this->current_pod);
+    //flow_solver->dg->solution = this->nearest_neighbors->nearestNeighborMidpointSolution(parameter);
```
```suggestion
```

biondicc replied:
removed  in new commit

cpethrick commented on src/testing/rom_import_helper_functions.cpp:
```diff
@@ -0,0 +1,120 @@
+#include <Epetra_Map.h>
+#include <Epetra_Vector.h>
+
+namespace PHiLiP {
```
Should this also be in the `Tests` namespace?

biondicc replied:
added to Tests namespace in .cpp and .h

cpethrick commented on src/reduced_order/hrom_test_location.h:
```diff
@@ -0,0 +1,81 @@
+    std::shared_ptr<DGBase<dim,double>> dg;
+
+    /// ECSW hyper-reduction weights
+    Epetra_Vector xi;
```
Can you change to `ECSW_weights` here too?

biondicc replied:
updated in new commit

biondicc commented on src/reduced_order/adaptive_sampling_base.cpp:
```diff
@@ -0,0 +1,352 @@
+        parameters.row(i) = snapshot_parameters.row(i);
+    }
+    this->pcout << i << std::endl;
+    // Loop through ROM points and add total errror to errors vector (both FOM snaps and ROM points are used to build RBF)
```
updated

cpethrick commented:
Everything looks good on my end! 

Great work on this PR, you've done a lot of good work with reorganizing the ROM portion of the code! I hope that your changes make it easier for future students to continue development of ROM/HROM :)

I will approve once you post the ctest results.

biondicc commented:
Here are the files which contain the tests that fail when I run ctest in my branch.

The first (Failed.log) is before I updated the helper functions, which were causing some of the hyperreduction tests to fail.
[Failed.log](https://github.com/user-attachments/files/15516836/Failed.log)

Once I updated the helper functions/some of the tolerances in my prm files in the latest commit, I re-ran ctest from test 274 onwards and the hyperreduction tests all passed, this was the new failed test log.
[LastTestsFailed.log](https://github.com/user-attachments/files/15516846/LastTestsFailed.log)

Thank you to the reviewers for all your helpful comments and support!
Let me know if there are still open questions.

cpethrick reviewed (approved):
Great work!
